leapp/reporting/__init__.py

```diff
@@ -272,6 +283,34 @@ def _sanitize_entries(entries):
         entries.append(Severity('info'))
     if not any(isinstance(e, Audience) for e in entries):
         entries.append(Audience('sysadmin'))
+    _check_stable_key(entries, raise_if_undefined=os.getenv('LEAPP_DEVEL_FIXED_REPORT_KEY', '0') != '0')
+
+
+def _check_stable_key(entries, raise_if_undefined=False):
+    """
+    Inserts a Key into the entries of the report if it doesn't exist.
+
+    The resulting id is calculated from the mandatory report
+    entries (severity, audience, title)
+    """
+    def _find_entry(cls):
+        entry = next((e for e in entries if isinstance(e, cls)), None)
+        if entry:
+            return entry._value
+
+    if _find_entry(Key):
+        # Key is already there, nothing to do
+        return
+    if raise_if_undefined:
+        raise ValueError('Stable Key report entry with specified unique value not provided')
+    else:
+        logger = configure_logger()
+        record_str = u'[{severity}:{audience}]{title}'.format(severity=_find_entry(Severity),
+                                                              audience=_find_entry(Audience),
+                                                              title=_find_entry(Title))
```

Comment:
As I remember from our tech.cabal meeting, we would like that all reports have a unique identifier, independent on the report content. Just any new report added to the codebase of some workflow (i.e. IPU) should have this key.

Then why do you generating a key. It is not added to the codebase and moreover, with changing let's say title of the report the key will become different.

Reply:
Changes to the actors will take place in leapp. 
Having dynamic fallback as a backup while having fixed ids set for all the report entries (in the follow up patches in the leapp-repository) and requiring all new code to have Key defined in the contributor documentation will give just the needed result.

Reply:
Our ideal state is all reports has a unique key. Leapp framework checks that all reports has it. What is the benefit to add unique keys in leapp-repos later, comparing to do it right now and merge this MR and MR in leapp-repos together. Moreover, I don't think there are a lot of work to add unique identifier to all lreports inside leapp-repos.
So I would go with:
1. Additional MR for leapp-repository, which adds unique ids to all reports (some script probably should be created)
2. Simplify this MR to the state, where it is just checking that all reports have a key, or raises an error
3. Merge these MRs together  
4. Require to rebase all open MRs in leapp-repository 

@pirat89 @vinzenz what do you think guys?

Reply:
My approach is a flawless gradual adoption of the feature that is even not there yet.

You are suggesting a make-everything-at-once change that will complicate life of every single leapp developer. Why?

Reply:
Maybe you're right, there are actors written in-house and they have own reports. In this case, a deprecation process may be applied to the reports.

Reply:
I think the only step we really need is 1/ @ZhukovGreen 

* Once this is merged, add explicit key to all currently existing calls of `create_report` in Leapp repository
  * All PRs will need to be rebased anyway, but I think we don't need to insist on rebasing everyhing *now*, only the PRs that are currently active
* Update our CI configuration to set `export LEAPP_DEVEL_FIXED_REPORT_KEY=1`

The problem with 2/ is that we're basically asking every 3rd party out there to refactor their actors else exceptions, and the automatic generation is nice middle ground.

@fernflower one thing I think we should solve in this PR is how are we going to treat report entries generated as a result of a dialog. IIRC report entries generated from dialogs have static title, audience and severity so they will all end up with the same digest. I think dialogs should also be able to set their key or autogenerate one from the actual question titles.

Reply:
@shaded-enmity about the dialogs - I'd say it is impossible to take care at the framework level, as currently the report about upgrade inhibition because of the unanswered dialog is happening in the leapp-repository https://github.com/oamg/leapp-repository/blob/master/repos/system_upgrade/el7toel8/actors/verifydialogs/libraries/verifydialogs.py#L23 

So this will be fixed by proper generation of the Key in the verifydialogs actor only, can't think of a nice way to embed this into the framework. What do you think?

Reply:
@pirat89 Thanks for pointing out, I will keep that in mind.
This PR however shouldn't affect anyone, it is the appropriate change in the actors and our guidelines, which, once merged, will require al yet-to-be-merged-actors to contain Key parameter.

Reply:
@fernflower you are right. That's caused by my fast reading-thinking today :) So far, until users will see any issues, it should be ok to merged without problems earlier than the other changes are implemented and require changes then.

Reply:
OK, then the origin of my concern is resolved. Resolving the comment